src/Server/RemoteFSHandler.cpp

```diff
@@ -0,0 +1,543 @@
+#include <Processors/Sinks/SinkToStorage.h>
+
+#include "Core/RemoteFSProtocol.h"
+#include "RemoteFSHandler.h"
+#include "Storages/MergeTree/RequestResponse.h"
+
+using namespace DB;
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ATTEMPT_TO_READ_AFTER_EOF;
+    extern const int UNKNOWN_PACKET_FROM_CLIENT;
+    extern const int UNKNOWN_DISK;
+    extern const int UNEXPECTED_PACKET_FROM_CLIENT;
+}
+
+
+RemoteFSHandler::RemoteFSHandler(
+    IServer & server_, TCPServer & tcp_server_, const Poco::Net::StreamSocket & socket_, std::string server_display_name_)
+    : Poco::Net::TCPServerConnection(socket_)
+    , server(server_)
+    , tcp_server(tcp_server_)
+    , log(&Poco::Logger::get("RemoteFSHandler"))
+    , server_display_name(std::move(server_display_name_))
+{
+}
+RemoteFSHandler::RemoteFSHandler(
+    IServer & server_,
+    TCPServer & tcp_server_,
+    const Poco::Net::StreamSocket & socket_,
+    TCPProtocolStackData & stack_data,
+    std::string server_display_name_)
+    : Poco::Net::TCPServerConnection(socket_)
+    , server(server_)
+    , tcp_server(tcp_server_)
+    , log(&Poco::Logger::get("RemoteFSHandler"))
+    , forwarded_for(stack_data.forwarded_for)
+    , certificate(stack_data.certificate)
+    , server_display_name(std::move(server_display_name_))
+{
+}
+
+RemoteFSHandler::~RemoteFSHandler()
+{
+    try
+    {
+        // TODO
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+void RemoteFSHandler::run()
+{
+    setThreadName("RemoteFSHandler");
+    ThreadStatus thread_status;
+
+    extractConnectionSettingsFromContext(server.context());
+
+    socket().setReceiveTimeout(receive_timeout);
+    socket().setSendTimeout(send_timeout);
+    socket().setNoDelay(true);
+
+    in = std::make_shared<ReadBufferFromPocoSocket>(socket());
+    out = std::make_shared<WriteBufferFromPocoSocket>(socket());
+
+    if (in->eof())
+    {
+        LOG_INFO(log, "Client has not sent any data.");
+        return;
+    }
+
+    try
+    {
+        receiveHello();
+        sendHello();
+    }
+    catch (const Exception & e)
+    {
+        if (e.code() == ErrorCodes::ATTEMPT_TO_READ_AFTER_EOF)
+        {
+            LOG_INFO(log, "Client has gone away.");
+            return;
+        }
+        if (e.code() == ErrorCodes::UNKNOWN_DISK)
+        {
+            LOG_TRACE(log, "Got error {}", e.message());
+            sendException(e);
+            return;
+        }
+        throw;
+    }
+
+    while (tcp_server.isOpen())
+    {
+        /// We are waiting for a packet from the client. Thus, every `poll_interval` seconds check whether we need to shut down.
+        {
+            Stopwatch idle_time;
+            UInt64 timeout_ms = std::min(poll_interval, idle_connection_timeout) * 1000000;
+            while (tcp_server.isOpen() && !server.isCancelled() && !static_cast<ReadBufferFromPocoSocket &>(*in).poll(timeout_ms))
```

Comment:
let's define `in` as `ReadBufferFromPocoSocket` in `RemoteFSHandler.h`, so there will be no need in static cast, same for `out`